File: ocaml/day2.ml
let read_input s =
  s 
  |> String.split_on_char 'x'
  |> List.map int_of_string
  |> List.sort compare
  |> List.rev;;

let task1 s =
  let sides_to_num s =
    let lens = read_input s in
    let prod = lens |> Utils.list_prod in 
    let main_areas = lens |> List.map (fun n -> prod / n) in
    let small_area = main_areas |> List.hd in

    2 * Utils.list_sum main_areas + small_area
  in

  s
  |> List.map sides_to_num
  |> Utils.list_sum
;;

let task2 s =
  let sides_to_num s =
    let lens = read_input s in
    let small_lens = lens |> List.tl in
    let prod = lens |> Utils.list_prod in 
    
    2 * Utils.list_sum small_lens + prod
  in

  s
  |> List.map sides_to_num
  |> Utils.list_sum
;;

let instructions = Utils.file_to_lines ("../input/day2.txt") in
let result1 = task1 instructions in
let result2 = task2 instructions in
  Printf.printf "Task1: %d\n" result1;
  Printf.printf "Task2: %d\n" result2;
  assert (result1 = 1586300);
  assert (result2 = 3737498);

File: cpp/day2.cpp
// Copyright 2023 Wojciech Sek

#include <cassert>
#include <numeric>
#include <string>
#include <vector>
#include <algorithm>

#include "utils.hpp"

int task1(const std::vector<std::vector<int>> &int_vectors) {
    int counter = 0;
    for (std::vector<int> int_vector : int_vectors) {
        int x = int_vector[0];
        int y = int_vector[1];
        int z = int_vector[2];
        int w = x * y;
        int h = y * z;
        int l = z * x;
        int small_area = std::min(w, std::min(h, l));

        counter +=  2 * (w + h + l) + small_area;
    }
    return counter;
}

int task2(const std::vector<std::vector<int>> &int_vectors) {
    int counter = 0;
    for (std::vector<int> int_vector : int_vectors) {
        int x = int_vector[0];
        int y = int_vector[1];
        int z = int_vector[2];
        int max = std::max(x, std::max(y, z));

        counter +=  2 * (x + y + z - max) + x * y * z;
    }
    return counter;
}

int main() {
    std::vector<std::vector<int>> input =
        file_to_int_vec2("../input/day2.txt", 'x');

    int result1 = task1(input);
    int result2 = task2(input);

    printf("Task1: %d\n", result1);
    printf("Task2: %d\n", result2);

    assert(result1 == 1586300);
    assert(result2 == 3737498);

    return 0;
}
